DIPs/DIP1NNN-CrazyPython.md

```diff
@@ -0,0 +1,283 @@
+| Status:         | Draft                                                           |
+## Abstract
+Often, it is necessary to allocate a data structure on the stack based on a
+size known at runtime, and possibly return that data structure. This proposal
+suggests adding dynamic-size static arrays to D. The goal here is an
+easy-to-use dynamically sized array without a GC. It should give the programmer
+the tools to avoid dynamic memory allocation.
+
+## Motivation
+`@nogc` programming is useful in many contexts. For example, currently WASM
+does not have a garbage collector, limiting D applications to `-betterC`
+mode. 
+
+It is often important to be able to quickly write `@nogc` code for
+performance-critical paths. After identifying hot functions using a profile,
+one should be able to alter a dynamic array that is only used within a single
+function to be allocated and deallocated automatically with `@nogc`.
+
+What is missing is an ergonomic way to dynamically allocate a static array on
+the stack built-in to D. You can take slices of static array as parameters
+
+```dlang
+/** A well-known algorithm for finding the topological sorting of a (possibly
+cyclic) directed graph. It deletes edges from its input as it runs, meaning
+adjacency_list is modified from within the function body.
+
+Because dynamic-size static arrays are value types, even when the function body
+modifies adjacency_list, the caller does not see the modifications.
+*/
+@nogc auto kahnsAlgorithm(int[][?] adjacency_list);
+```
+
+
+## Definition and key features
+
+<!-- rephrase: use the word "arguments" for inside the body, "parameters" for
+outside it -->
+<!-- Todo: add unit tests -->
+<!-- Todo: write Kahn's Algorithm in D and showcase the various solutions -->
+
+Like normal D static arrays, dynamic-size static arrays are pass-by-value: 
+
+```dlang
+@nogc void no_op(int[?] list) {
+for (int i = 0; i < list.length; ++i) {
+list[i] = 0;
+}
+}
+unittest {
+int[5] array = [1, 2, 3, 4, 5];
+no_op(array);
+assert(array == [1, 2, 3, 4, 5]);
+}
+```
+
+A dynamic array can be implicitly converted to a dynamic-size static array as a
+copy:
+
+```
+unittest {
+int[] array = [1, 2, 3, 4, 5];
+no_op(array);
+assert(array == [1, 2, 3, 4, 5]);
+int[?] another_array = array;
+another_array[0] = 999;
+assert(array[0] != 999);
+}
+unittest {
+int[][] array = [[1, 2], [3, 4], [5, 6]];
+int[?][?] secondarray = array;
+}
+```
+
+Dynamic-size static arrays have a `.length` property. The `.length` property
+may be changed arbitrarily, just like for a dynamic array. 
+
+**Implementation Defined:** Enough space to hold the original length of the
+array is allocated on the stack.
+
+**Implementation Defined:** Shrinking a dynamic-size static array does not free
+or deallocate the memory held by a dynamic-size static array.
+
+**Implementation Defined:** Expanding a dynamic-sized array's `.length`
+property beyond its *original length* may allocate memory on the heap. 
```

Comment:
The compiler will be unable to statically infer when the program may resize the array beyond its original length. This means that it will have to link in the GC unconditionally.

Reply:
In some code, the C++ compiler does not always know when a `std::vector<int>` will be resized the array beyond its original length. Yet, it does not need a GC.

Reply:
You can return a dynamic-size static array, but you can never return a slice of one. Shrinking a dynamic size static array does not deallocate any memory.

I think that these two conditions ensure the essential memory safety of this feature, with or without GC.